File: apps/demo-be-csharp-aspnetcore/tests/DemoBeCsas.Tests/Integration/Steps/ExpenseSteps.cs
using System.Text.Json;
using DemoBeCsas.Tests.ScenarioContext;
using FluentAssertions;
using Reqnroll;
using Xunit;

namespace DemoBeCsas.Tests.Integration.Steps;

[Binding]
[Trait("Category", "Integration")]
public class ExpenseSteps(ServiceLayer svc, SharedState state, AuthSteps auth)
{
    internal Guid? _bobExpenseId;

    // ─────────────────────────────────────────────────────────────
    // Given helpers
    // ─────────────────────────────────────────────────────────────

    [Given(@"^alice has created an entry with body (.+)$")]
    public async Task GivenAliceCreatedEntry(string bodyJson)
    {
        var (desc, title, category, amount, currency, type, quantity, unit, date) =
            ParseExpenseBody(bodyJson);
        var response = await svc.CreateExpenseAsync(
            state.AccessToken,
            desc,
            title,
            category,
            amount,
            currency,
            type,
            quantity,
            unit,
            date
        );
        ((int)response.StatusCode).Should().Be(
            201,
            $"Failed to create entry: {response.Body}"
        );
        using var doc = JsonDocument.Parse(response.Body);
        if (doc.RootElement.TryGetProperty("id", out var idEl))
        {
            state.LastExpenseId = Guid.Parse(idEl.GetString()!);
        }
    }

    [Given(@"^alice has created an expense with body (.+)$")]
    public async Task GivenAliceCreatedExpense(string bodyJson) =>
        await GivenAliceCreatedEntry(bodyJson);

    [Given(@"^alice has created 3 entries$")]
    public async Task GivenAliceCreated3Entries()
    {
        var bodies = new[]
        {
            """{ "amount": "10.00", "currency": "USD", "category": "food", "description": "Entry 1", "date": "2025-01-01", "type": "expense" }""",
            """{ "amount": "20.00", "currency": "USD", "category": "food", "description": "Entry 2", "date": "2025-01-02", "type": "expense" }""",
            """{ "amount": "30.00", "currency": "USD", "category": "food", "description": "Entry 3", "date": "2025-01-03", "type": "expense" }""",
        };
        foreach (var body in bodies)
        {
            await GivenAliceCreatedEntry(body);
        }
    }

    [Given(@"^bob has created an entry with body (.+)$")]
    public async Task GivenBobCreatedEntry(string bodyJson)
    {
        var (bobToken, _) = await auth.LoginUserAsync("bob", "Str0ng#Pass2");
        bobToken.Should().NotBeNull("bob should be able to login");

        var (desc, title, category, amount, currency, type, quantity, unit, date) =
            ParseExpenseBody(bodyJson);
        var response = await svc.CreateExpenseAsync(
            bobToken,
            desc,
            title,
            category,
            amount,
            currency,
            type,
            quantity,
            unit,
            date
        );
        ((int)response.StatusCode).Should().Be(
            201,
            $"Failed to create bob's entry: {response.Body}"
        );
        using var doc = JsonDocument.Parse(response.Body);
        if (doc.RootElement.TryGetProperty("id", out var idEl))
        {
            _bobExpenseId = Guid.Parse(idEl.GetString()!);
        }
    }

    // ─────────────────────────────────────────────────────────────
    // When steps
    // ─────────────────────────────────────────────────────────────

    [When(@"^alice sends POST /api/v1/expenses with body (.+)$")]
    public async Task WhenAliceCreatesExpense(string bodyJson)
    {
        var (desc, title, category, amount, currency, type, quantity, unit, date) =
            ParseExpenseBody(bodyJson);
        state.LastResponse = await svc.CreateExpenseAsync(
            state.AccessToken,
            desc,
            title,
            category,
            amount,
            currency,
            type,
            quantity,
            unit,
            date
        );
        if (state.LastResponse.IsSuccess)
        {
            using var doc = JsonDocument.Parse(state.LastResponse.Body);
            if (doc.RootElement.TryGetProperty("id", out var idEl))
            {
                state.LastExpenseId = Guid.Parse(idEl.GetString()!);
            }
        }
    }

    [When(@"^the client sends POST /api/v1/expenses with body (.+)$")]
    public async Task WhenUnauthenticatedClientCreatesExpense(string bodyJson)
    {
        var (desc, title, category, amount, currency, type, quantity, unit, date) =
            ParseExpenseBody(bodyJson);
        state.LastResponse = await svc.CreateExpenseAsync(
            null, // no token — unauthenticated
            desc,
            title,
            category,
            amount,
            currency,
            type,
            quantity,
            unit,
            date
        );
    }

    [When(@"^alice sends GET /api/v1/expenses/\{expenseId\}$")]
    public async Task WhenAliceGetsExpense()
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        state.LastResponse = await svc.GetExpenseAsync(state.AccessToken, state.LastExpenseId!.Value);
    }

    [When(@"^alice sends GET /api/v1/expenses$")]
    public async Task WhenAliceListsExpenses()
    {
        state.LastResponse = await svc.ListExpensesAsync(state.AccessToken);
    }

    [When(@"^alice sends PUT /api/v1/expenses/\{expenseId\} with body (.+)$")]
    public async Task WhenAliceUpdatesExpense(string bodyJson)
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        var (desc, title, category, amount, currency, type, quantity, unit, date) =
            ParseExpenseBody(bodyJson);
        state.LastResponse = await svc.UpdateExpenseAsync(
            state.AccessToken,
            state.LastExpenseId!.Value,
            desc,
            title,
            category,
            amount,
            currency,
            type,
            quantity,
            unit,
            date
        );
    }

    [When(@"^alice sends DELETE /api/v1/expenses/\{expenseId\}$")]
    public async Task WhenAliceDeletesExpense()
    {
        state.LastExpenseId.Should().NotBeNull("expense ID should be set");
        state.LastResponse = await svc.DeleteExpenseAsync(
            state.AccessToken,
            state.LastExpenseId!.Value
        );
    }

    [When(@"^alice sends GET /api/v1/expenses/summary$")]
    public async Task WhenAliceGetsSummary()
    {
        state.LastResponse = await svc.GetExpenseSummaryAsync(state.AccessToken);
    }

    // ─────────────────────────────────────────────────────────────
    // Then steps — numeric field comparisons
    // ─────────────────────────────────────────────────────────────

    [Then(@"^the response body should contain ""([^""]+)"" equal to (\d+(?:\.\d+)?)$")]
    public void ThenResponseBodyContainsNumericField(string field, double value)
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        using var doc = JsonDocument.Parse(body);
        doc.RootElement.TryGetProperty(field, out var el)
            .Should().BeTrue($"Field '{field}' not found in: {body}");
        var actual = el.ValueKind == JsonValueKind.Number
            ? el.GetDouble()
            : double.Parse(el.GetString()!);
        actual.Should().BeApproximately(value, 0.0001);
    }

    [Then(@"^the response body should contain ""([^""]+)"" total equal to ""([^""]+)""$")]
    public void ThenResponseBodyContainsCurrencyTotal(string currency, string expectedTotal)
    {
        state.LastResponse.Should().NotBeNull();
        var body = state.LastResponse!.Body;
        using var doc = JsonDocument.Parse(body);
        doc.RootElement.TryGetProperty(currency, out var el)
            .Should().BeTrue($"Currency '{currency}' not found in: {body}");
        var actual = el.ValueKind == JsonValueKind.Number
            ? el.GetDecimal()
            : decimal.Parse(el.GetString()!);
        actual.Should().Be(decimal.Parse(expectedTotal));
    }

    // ─────────────────────────────────────────────────────────────
    // Helpers
    // ─────────────────────────────────────────────────────────────

    private static (
        string? Description,
        string? Title,
        string? Category,
        decimal Amount,
        string? Currency,
        string? Type,
        double? Quantity,
        string? Unit,
        string? Date
    ) ParseExpenseBody(string bodyJson)
    {
        using var doc = JsonDocument.Parse(bodyJson);
        var root = doc.RootElement;

        string? description = root.TryGetProperty("description", out var d) ? d.GetString() : null;
        string? title = root.TryGetProperty("title", out var t) ? t.GetString() : null;
        string? category = root.TryGetProperty("category", out var c) ? c.GetString() : null;
        string? currency = root.TryGetProperty("currency", out var cur) ? cur.GetString() : null;
        string? type = root.TryGetProperty("type", out var tp) ? tp.GetString() : null;
        string? unit = root.TryGetProperty("unit", out var u) ? u.GetString() : null;
        string? date = root.TryGetProperty("date", out var dt) ? dt.GetString() : null;

        decimal amount = 0m;
        if (root.TryGetProperty("amount", out var amtEl))
        {
            if (amtEl.ValueKind == JsonValueKind.Number)
            {
                amount = amtEl.GetDecimal();
            }
            else if (amtEl.ValueKind == JsonValueKind.String)
            {
                amount = decimal.Parse(amtEl.GetString()!, System.Globalization.CultureInfo.InvariantCulture);
            }
        }

        double? quantity = null;
        if (root.TryGetProperty("quantity", out var qEl) && qEl.ValueKind == JsonValueKind.Number)
        {
            quantity = qEl.GetDouble();
        }

        return (description, title, category, amount, currency, type, quantity, unit, date);
    }
}

File: apps/demo-be-python-fastapi/tests/unit/steps/expense_steps.py
"""BDD step definitions for expense management feature."""

import json

from pytest_bdd import given, parsers, scenarios, when

from tests.integration.service_client import FakeResponse, ServiceClient
from tests.unit.conftest import GHERKIN_ROOT

scenarios(str(GHERKIN_ROOT / "expenses" / "expense-management.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token',
    target_fixture="alice_tokens",
)
def alice_login_expense(client: ServiceClient, registered_user: dict) -> dict:
    return client.login_user("alice", _PASSWORD)


@given(
    parsers.parse("alice has created an entry with body {body}"),
    target_fixture="created_expense",
)
def alice_create_entry(client: ServiceClient, alice_tokens: dict, body: str) -> dict:
    data = json.loads(body)
    resp = client.post_expense(f"Bearer {alice_tokens['accessToken']}", data)
    assert resp.status_code == 201, f"Create expense failed: {resp.text}"
    return resp.json()


@given("alice has created 3 entries", target_fixture="created_expenses")
def alice_create_3_entries(client: ServiceClient, alice_tokens: dict) -> list:
    expenses = []
    for i in range(3):
        resp = client.post_expense(
            f"Bearer {alice_tokens['accessToken']}",
            {
                "amount": f"{10 + i}.00",
                "currency": "USD",
                "category": "food",
                "description": f"Entry {i}",
                "date": f"2025-01-{10 + i:02d}",
                "type": "expense",
            },
        )
        assert resp.status_code == 201
        expenses.append(resp.json())
    return expenses


# --- When steps ---


@when(
    parsers.parse("alice sends POST /api/v1/expenses with body {body}"),
    target_fixture="response",
)
def alice_post_expense(client: ServiceClient, alice_tokens: dict, body: str) -> FakeResponse:
    data = json.loads(body)
    return client.post_expense(f"Bearer {alice_tokens['accessToken']}", data)


@when(
    'the client sends POST /api/v1/expenses with body { "amount": "10.00", "currency": "USD", "category": "food", "description": "Coffee", "date": "2025-01-01", "type": "expense" }',  # noqa: E501
    target_fixture="response",
)
def unauth_post_expense(client: ServiceClient) -> FakeResponse:
    return client.post_expense(
        None,
        {
            "amount": "10.00",
            "currency": "USD",
            "category": "food",
            "description": "Coffee",
            "date": "2025-01-01",
            "type": "expense",
        },
    )


@when("alice sends GET /api/v1/expenses/{expenseId}", target_fixture="response")
def alice_get_expense(
    client: ServiceClient, alice_tokens: dict, created_expense: dict
) -> FakeResponse:
    return client.get_expense(
        created_expense["id"],
        f"Bearer {alice_tokens['accessToken']}",
    )


@when("alice sends GET /api/v1/expenses", target_fixture="response")
def alice_list_expenses(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.get_expenses(f"Bearer {alice_tokens['accessToken']}")


@when(
    parsers.parse("alice sends PUT /api/v1/expenses/{{expenseId}} with body {body}"),
    target_fixture="response",
)
def alice_update_expense(
    client: ServiceClient, alice_tokens: dict, created_expense: dict, body: str
) -> FakeResponse:
    data = json.loads(body)
    return client.put_expense(
        created_expense["id"],
        f"Bearer {alice_tokens['accessToken']}",
        data,
    )


@when("alice sends DELETE /api/v1/expenses/{expenseId}", target_fixture="response")
def alice_delete_expense(
    client: ServiceClient, alice_tokens: dict, created_expense: dict
) -> FakeResponse:
    return client.delete_expense(
        created_expense["id"],
        f"Bearer {alice_tokens['accessToken']}",
    )
